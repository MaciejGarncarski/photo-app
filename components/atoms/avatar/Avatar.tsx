import { IconUser } from '@tabler/icons';
import clsx from 'clsx';
import { motion } from 'framer-motion';
import Image, { ImageProps } from 'next/image';

import { useUser } from '@/hooks/useUser';

import { VisuallyHiddenText } from '@/components/atoms/visuallyHiddenText/VisuallyHiddenText';

import styles from './avatar.module.scss';

type PropsTypes = Partial<Pick<ImageProps, 'width' | 'height'>> & {
  userId?: string;
  className?: string;
};

const DEFAULT_AVATAR_SIZE = 140;

export const MotionImage = motion(Image);

export const Avatar = ({
  userId,
  className,
  width = DEFAULT_AVATAR_SIZE,
  height = DEFAULT_AVATAR_SIZE,
}: PropsTypes) => {
  const { customImage, image, username } = useUser({ userId });

  const hasNoImage = Boolean(!image && !customImage);
  const hasDefaultImage = image && !Boolean(customImage);
  const hasCustomImage = customImage;

  return (
    <figure className={clsx(className, styles.avatarContainer)}>
      {hasNoImage && (
        <div data-testid="empty icon">
          <IconUser className={styles.icon} />
          <VisuallyHiddenText text={username ?? ''} />
        </div>
      )}
      {hasDefaultImage && (
        <MotionImage src={image} alt={username ?? ''} width={width} height={height} className={styles.avatar} />
      )}
      {hasCustomImage && (
        <MotionImage
          src={customImage}
          alt={username ?? ''}
          data-testid="customImage"
          width={width}
          height={height}
          className={styles.avatar}
        />
      )}
    </figure>
  );
};
