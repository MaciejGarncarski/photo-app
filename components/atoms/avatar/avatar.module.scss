.avatar {
  border-radius: 100%;
}