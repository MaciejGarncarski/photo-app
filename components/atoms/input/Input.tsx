import clsx from 'clsx';
import { ChangeEvent, forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type PropsTypes = {
  type?: 'text' | 'number' | 'tel' | 'email' | 'password';
  labelText: string;
  error?: FieldError;
  optional?: boolean;
  className?: string;
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
  value?: string;
};

export const Input = forwardRef<HTMLInputElement, PropsTypes>(
  ({ type = 'text', labelText, error, optional, className, onChange, value, ...props }, ref) => {
    const containerClassName = clsx(className, error && styles.containerError, styles.container);

    const inputClassName = clsx(
      {
        [styles.inputError]: Boolean(error),
      },
      styles.input,
    );

    return (
      <div className={containerClassName}>
        <label className={styles.label} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
        <input
          ref={ref}
          type={type}
          id={labelText}
          data-testid="input"
          className={inputClassName}
          onChange={onChange}
          value={value}
          {...props}
        />
        {error && <p className={styles.error}>{error.message}</p>}
      </div>
    );
  },
);
