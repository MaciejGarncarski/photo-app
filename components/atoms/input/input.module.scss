@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--text-secondary);
  text-transform: capitalize;
}


.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
  border: var(--border-gray);
  border-radius: var(--radius-medium);
  transition: all var(--transition);
  transition-property: background-color, box-shadow;

  &:hover {
    background-color: var(--bg-secondary);
  }

  &:focus {
    outline: none;
    border-color: transparent;
    box-shadow: var(--shadow-focus);
  }
}

.inputError {
  background-color: var(--bg-error);
  border-color: var(--color-error);
}


.error {
  color: var(--color-error);
  font-weight: 600;
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  resize: vertical;
}