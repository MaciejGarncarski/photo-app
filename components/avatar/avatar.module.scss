@use 'common';

.avatar {
  --avatar-size: 2rem;
  width: var(--avatar-size);
  height: var(--avatar-size);
  border-radius: 100%;
  --border-size: 2px;
  padding: var(--border-size);
  display: flex;
  justify-content: center;
  align-items: center;
  vertical-align: middle;
  position: relative;
  overflow: hidden;
  flex-shrink: 0;
  z-index: 0;
  box-sizing: border-box;

  &::before {
    content: '';
    width: 100%;
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;
    z-index: -1;
    border-radius: 999px;
    background-image: linear-gradient(
      135deg,
      var(--color-accent),
      var(--color-accent-darker)
    );
  }

  img {
    background-color: var(--color-primary);
    object-fit: cover;
    display: flex;
    width: 100%;
    height: 100%;
    border-radius: 999px;
  }
}

.small {
  --avatar-size: 2.2rem;

  @include common.mediaquery('desktop') {
    --avatar-size: 2.5rem;
  }
}

.medium {
  --avatar-size: 3.5rem;

  @include common.mediaquery('desktop') {
    --avatar-size: 4.15rem;
  }
}

.big {
  --avatar-size: 5rem;
  --ring-size: 5px;

  @include common.mediaquery('desktop') {
    --avatar-size: 8rem;
  }
}

.noImage {
  width: 100%;
  height: 100%;
  display: flex;
  border-radius: 100%;
  justify-content: center;
  align-items: center;
  background-color: var(--color-primary);
}

.imagePlaceholder {
  width: 60%;
  margin: auto;
  height: auto;
}
