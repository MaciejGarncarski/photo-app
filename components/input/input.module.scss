@use 'common';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
  --gap: 0.7rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--color-text-gray);
  transition: all var(--transition);
  transition-property: background-color, transform, opacity;
  display: inline-block;
}

.input {
  background-color: var(--color-secondary);
  padding: 1rem var(--gap);
  border-radius: var(--radius-small);
  border: 1px solid var(--shadow-0-5);
  box-shadow: var(--shadow);
  color: var(--color-text);

  &:focus {
    outline: none;
  }
}

.primary {
  background-color: var(--color-secondary);
}

.secondary {
  @include common.mediaquery('desktop') {
    background-color: var(--color-primary);
  }
}

.inputError {
  background-color: var(--color-error-light);

  &::placeholder {
    color: var(--color-text);
    font-weight: 500;
  }
}

.error {
  color: var(--color-error);
  padding-left: var(--gap);
  font-weight: 400;
  margin-left: 0.75rem;
  margin-top: 0.25rem;
}
