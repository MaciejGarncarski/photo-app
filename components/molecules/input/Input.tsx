import clsx from 'clsx';
import { forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type InputProps = {
  type?: JSX.IntrinsicElements['input']['type'];
  labelText: string;
  isDirty?: boolean;
  error?: FieldError;
  optional?: boolean;
  className?: string;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ type = 'text', labelText, isDirty, error, optional, className, ...props }, ref) => {
    return (
      <div
        className={clsx(className && className, error && styles.containerError, styles.container)}
      >
        <input className={styles.input} ref={ref} type={type} {...props} id={labelText} />
        <label className={clsx(isDirty && styles.labelActive, styles.label)} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
      </div>
    );
  }
);
