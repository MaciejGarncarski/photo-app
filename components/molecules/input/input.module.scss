@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  border: 2px solid var(--color-primary);
  border-radius: var(--radius-small);
  position: relative;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  position: absolute;
  left: 0rem;
  top: 0;
  padding: 0 1rem;
  color: var(--text-placeholder);
  transition: all 100ms ease-out;
  cursor: text;
  transition-property: background-color, transform, font-size;
}

.labelActive,
.input:focus+.label {

  background-color: var(--bg-primary);
  color: var(--text-primary);
  top: -1.25rem;
  left: 0.5rem;


  @include common.mediaquery('lg') {
    font-size: 2.5rem;
  }

  @include common.mediaquery('xl') {
    top: -2rem;
    left: 1rem;
  }
}

.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
}

.error {
  color: var(--color-error);
  font-weight: 600;
  font-size: 1.5rem;
  margin-left: 1rem;

  @include common.mediaquery('lg') {
    font-size: 2rem;
  }
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  font-size: 2rem;
  resize: vertical;
}