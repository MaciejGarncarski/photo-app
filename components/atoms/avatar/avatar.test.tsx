import { screen, waitFor } from '@testing-library/react';

import { mockedUser } from '@/utils/tests/mockedData';
import { render } from '@/utils/tests/utils';

import { Avatar } from '@/components/atoms/avatar/Avatar';
import { snapshotComponent } from '@/components/molecules/input/input.test';

jest.mock('next-auth/react', () => {
  const originalModule = jest.requireActual('next-auth/react');
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { name: mockedUser.user.name, username: mockedUser.user.username },
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: jest.fn(() => {
      return { data: mockSession, status: 'authenticated' };
    }),
  };
});

describe('Avatar', () => {
  describe('match snapshot', () => {
    snapshotComponent(<Avatar />);
  });

  it('Should render avatar with no image.', () => {
    render(<Avatar />);
    expect(screen.getByTestId('empty icon')).toBeInTheDocument();
  });

  it('Should render users avatar', async () => {
    render(<Avatar userId="user" />);
    await waitFor(() => expect(screen.getByTestId('customImage')).toBeInTheDocument(), { timeout: 2000 });
  });
});
