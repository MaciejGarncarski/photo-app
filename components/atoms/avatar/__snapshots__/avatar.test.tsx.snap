// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <label
      class="label"
      for="random"
    >
      random 
    </label>
    <input
      class="input"
      data-testid="test"
      id="random"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[` 2`] = `
<DocumentFragment>
  <figure
    class="avatar"
  >
    <div
      class="noImage"
      data-testid="empty icon"
    >
      <svg
        class="icon icon-tabler icon-tabler-user"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
          stroke="none"
        />
        <circle
          cx="12"
          cy="7"
          r="4"
        />
        <path
          d="M6 21v-2a4 4 0 0 1 4 -4h4a4 4 0 0 1 4 4v2"
        />
      </svg>
      <span
        class="visually-hidden"
      />
    </div>
  </figure>
</DocumentFragment>
`;
